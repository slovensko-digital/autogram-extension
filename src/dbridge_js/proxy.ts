const log = [];

const ignoredProps: PropertyKey[] = [
  "__implementation",
  Symbol.toStringTag,
  Symbol.iterator,
  "$$typeof",
  "@@__IMMUTABLE_RECORD__@@",
  "hasAttribute",
  "tagName",
  "nodeType",
  "onSuccess",
  "toString",
  Symbol.toPrimitive,
  Symbol.toStringTag,
  "splice",
  "prototype",
];

const logger = {
  push(...rest) {
    // console.log(...rest);
    log.push(...rest);
  },
};

function getSpyHandler(prefix?: string, depth?: number) {
  const spyHandler = {
    get(target: object, prop: PropertyKey, receiver?: unknown) {
      //   log.push({ target, prefix, prop, receiver });
      //   const wrappedTarget =
      //     typeof target === "function" ? wrapFunction(target) : target;

      if (prop === "__proxy_name") {
        return prefix;
      }

      if (prop === "__proxy_log") {
        return log;
      }

      const name = `${prefix}.${prop.toString()}`;
      const reflection = Reflect.get(target, prop, receiver);

      if (typeof reflection == "function")
        logger.push({
          type: "get",
          name,
          target,
          prop,
          receiver,
        });

      return depth < 3 &&
        !ignoredProps.includes(prop) &&
        (typeof reflection == "function" || typeof reflection == "object")
        ? wrapWithProxy(reflection, name, depth + 1)
        : reflection;
    },

    // eslint-disable-next-line @typescript-eslint/ban-types
    // construct(target: Function, argumentList, newTarget: Function) {
    //   console.log("construct", target, argumentList, newTarget);
    //   return Reflect.construct(target, argumentList, newTarget);
    // },

    // eslint-disable-next-line @typescript-eslint/ban-types
    apply(target: Function, thisArg, argumentList) {
      const name = `${thisArg.__proxy_name}.${target.name}`;

      // this causes infinite loop
      if (
        argumentList &&
        argumentList[0] &&
        argumentList[0]["onSuccess"] &&
        typeof argumentList[0]["onSuccess"] === "function"
      ) {
        const oldOnSuccess = argumentList[0]["onSuccess"];
        argumentList[0]["onSuccess"] = function (...onSuccessArgs) {
          logger.push({ type: "callback.onSuccess", name, onSuccessArgs });
          return oldOnSuccess(...onSuccessArgs);
        };
      }

      logger.push({ type: "apply", name, target, thisArg, argumentList });
      return Reflect.apply(target, thisArg, argumentList);
    },
  };
  return spyHandler;
}

export function wrapWithProxy<T extends object>(
  target: T,
  prefix = "root",
  depth = 0
): T & { __proxy_log: typeof log } {
  if (typeof target !== "object" && typeof target !== "function") {
    // console.log("skipping", typeof target, target);
    return target;
  }
  return new Proxy(target, getSpyHandler(prefix, depth)) as T & {
    __proxy_log: typeof log;
  };
}
