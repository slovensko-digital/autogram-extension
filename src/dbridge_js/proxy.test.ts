import { wrapWithProxy } from "./proxy";

describe("proxy", () => {
  test("to run", () => {
    const target = {
      a: "test",
      b: { a: "test" },
      fn: (callback) => {
        target.a = "x";
        callback.onSuccess("success");
      },
    };
    const wrapped = wrapWithProxy(target);
    wrapped.a;
    wrapped.fn({
      onSuccess: (response) => {
        console.log(response);
      },
    });
    wrapped.toString()
    console.log(wrapped.__proxy_log);
    expect(wrapped.__proxy_log).toBe([
      {
        type: "get",
        target: { a: "x", b: null, fn: null },
        prop: "a",
        receiver: { a: "x", b: null, fn: null },
      },
      {
        type: "get",
        target: { a: "x", b: null, fn: null },
        prop: "fn",
        receiver: { a: "x", b: null, fn: null },
      },
      {
        type: "apply",
        target: null,
        thisArg: { a: "x", b: null, fn: null },
        name: "root.fn",
        argumentList: [],
      },
    ]);
  });
});
