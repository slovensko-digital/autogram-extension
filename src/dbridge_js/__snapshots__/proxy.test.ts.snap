// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`proxy to run 1`] = `
Array [
  Object {
    "prop": "a",
    "receiver": Object {
      "a": "test",
      "b": Object {
        "a": "test",
      },
      "fn": [Function],
    },
    "target": Object {
      "a": "test",
      "b": Object {
        "a": "test",
      },
      "fn": [Function],
    },
    "type": "get",
  },
]
`;
